designs/NuGet-ClientCertificates.md

```diff
@@ -0,0 +1,114 @@
+Equals '**FindByThumbprint**' by default
+- Optional Key="**FindValue**" Value="The search criteria as an object"
+
+- **fromCert** - item for certificate import directly from file (DER encoded x.509, Base-64 encoded x.509, PKCS)
+Can be configured with 2 Add item's and item text body
+
+- Key="**Path**" Value="Absolute or relative path to certificate file". Relative path resolve have 2 stages: first relative to configuration file origin, second relative to current application directory.
+- Optional Key="**Password**" Value="Encrypted password". Password for certificate file. Encrypted in same manner as PackageSourceCredential password.
+- Items body can be filled with Base-64 encoded x.509 certificate. Start and end of text are trimmed from spaces and new line chars.
+
+## Configuration example
+
+```xml
+<configuration>
+...
+    <clientCertificates>	
+        <fromStorage>
+            <!-- Optional. CurrentUser by default -->
+            <add key="StoreLocation" value="CurrentUser" />
+            <!-- Optional. My by default -->
+            <add key="StoreName" value="My" />
+            <!-- Optional. FindByThumbprint by default -->
+            <add key="FindType" value="FindByThumbprint" />
+            <!-- Required. -->
+            <add key="FindValue" value="4894671ae5aa84840cc1079e89e82d426bc24ec6" />
+        </fromStorage>
+        <!-- Form 1 -->
+        <fromCert>
+            <!-- Absolute or relative path to certificate file. -->
+            <add key="Path" value=".\certificate.pfx" />
+            <!-- Encrypted password -->
+            <add key="Password" value="..." />
+        </fromCert>
+        <!-- Form 2 -->
+        <fromCert>
+-----BEGIN CERTIFICATE-----
+MIIEjjCCA3agAwIBAgIJIBkBGf8AAAAPMA0GCSqGSIb3DQEBCwUAMIGzMQswCQYD
+...
+tJl1UvF7GWJd0yNyPVqCCnBY
+-----END CERTIFICATE-----
+        </fromCert>
+    </clientCertificates>
+...
+</configuration>
+```
+
+## Configuration components implementation
+
+Added several new SettingItem's to parse configuration above. Certificates search logic encapsulated inside.
+
+For high level usage were introduced:
+
+`ClientCertificateProvider` class
+* `Provide(settings:ISettings) : IEnumerable<X509Certificate>` - extracts certificates from settings
+
+`ClientCertificates` static singletone certificate storage.
+* `Store(certificates: IEnumerable<X509Certificate>)` - stores certificate instances
+* `SetupClientHandler(httpClientHandler:HttpClientHandler)` - adds stored certificates to HttpClientHandler instance. 
+
+## NuGet Cli implementation as example
+
+1) Internally certificates extracted from configuration items on base [Command.Execute](https://github.com/NuGet/NuGet.Client/blob/d4f53c3e523493fcbe35c537cb004e9a3e228abd/src/NuGet.Clients/NuGet.CommandLine/Commands/Command.cs) with `ClientCertificateProvider.Provide` method
```

Comment:
again, the spec is supposed to be the "source of truth", so if there's a bug, or if someone decides to rewrite the client from scratch, the specs should contain the information required. The spec shouldn't refer back to implementations, or PRs, to specify how the feature works.

Reply:
Plaese, look at new version. Is it better?